Remove telemetry, SmartDashboard, and indirect entrypoint functionality

al-rigazzi opened the pull request:
## Overview

This PR removes telemetry collection, SmartDashboard functionality, and indirect entrypoint capabilities from SmartSim. Additionally, it significantly improves the codebase's maintainability by centralizing directory path management through the CONFIG system and eliminating hardcoded `.smartsim` directory references throughout the codebase.

## Changes Made

### 🗑️ Removed Features
- **Telemetry Collection**: Removed all telemetry data collection and transmission functionality
- **SmartDashboard**: Removed dashboard visualization and monitoring capabilities  
- **Indirect Entrypoints**: Removed support for indirect application entrypoints

### 🏗️ Configuration System Improvements
- **Centralized Directory Management**: Enhanced the CONFIG system with comprehensive directory path properties
- **Hierarchical Structure**: Implemented proper directory hierarchy for better organization
- **Eliminated Hardcoded Paths**: Replaced all hardcoded `.smartsim` directory references with CONFIG properties

### 📁 New CONFIG Properties
```python
CONFIG.smartsim_base_dir        # Returns ".smartsim"
CONFIG.dragon_default_subdir   # Returns ".smartsim/dragon"  
CONFIG.dragon_logs_subdir       # Returns ".smartsim/dragon/logs"
CONFIG.metadata_subdir          # Returns ".smartsim/metadata"
```

### 📂 Directory Structure
The new hierarchical structure provides better organization:
```
.smartsim/
├── dragon/
│   └── logs/           # Dragon launcher logs
├── metadata/           # Experiment metadata
│   └── run_{timestamp}/
│       ├── model/
│       ├── ensemble/
│       └── database/
└── keys/              # Security keys
```

## Files Modified

### Core Configuration
- `smartsim/_core/config/config.py` - Enhanced with new directory properties and hierarchical structure

### Core Functionality  
- `smartsim/_core/utils/serialize.py` - Updated to use CONFIG.metadata_subdir
- `smartsim/_core/control/manifest.py` - Updated to use CONFIG.metadata_subdir

### Test Files (15+ files updated)
**High Priority:**
- `tests/test_symlinking.py` - Updated hardcoded paths to use CONFIG properties
- `tests/test_manifest_metadata_directories.py` - Updated to use CONFIG.metadata_subdir
- `tests/test_metadata_integration.py` - Updated to use CONFIG properties

**Medium Priority:**
- `tests/test_controller.py` - Updated to use CONFIG.metadata_subdir
- `tests/test_output_files.py` - Updated comments to reference CONFIG

**Dragon Test Files:**
- `tests/test_dragon_step.py` - Updated to use CONFIG.dragon_logs_subdir
- `tests/test_dragon_launcher.py` - Updated to use CONFIG.dragon_logs_subdir  
- `tests/test_dragon_client.py` - Updated to use CONFIG.dragon_logs_subdir
- `tests/test_dragon_run_policy.py` - Updated to use CONFIG.dragon_logs_subdir

## Benefits

### 🔧 Improved Maintainability
- **Single Source of Truth**: All directory paths now managed through CONFIG
- **Easy Customization**: Directory structure can be modified by changing CONFIG properties
- **Consistent Architecture**: Eliminates path duplication and inconsistencies

### 🧪 Better Testing
- **Configurable Paths**: Tests can easily modify directory paths through CONFIG
- **Consistent Test Environment**: All tests use the same path management system
- **Future-Proof**: New directory requirements can be added as CONFIG properties

### 🏗️ Enhanced Modularity
- **Hierarchical Organization**: Dragon components properly organized under parent directory
- **Logical Structure**: Related functionality grouped in appropriate subdirectories
- **Extensible Design**: Easy to add new directory categories

## Technical Details

### CONFIG System Architecture
The new CONFIG properties follow a hierarchical dependency model:
- `smartsim_base_dir` is the foundation (`.smartsim`)
- `dragon_default_subdir` builds on base (`{base}/dragon`)
- `dragon_logs_subdir` builds on dragon (`{dragon}/logs`)
- `metadata_subdir` builds on base (`{base}/metadata`)

This ensures that changing the base directory automatically propagates to all subdirectories, and changing the dragon directory affects the dragon logs directory.

### Path Management Implementation
All directory references now use f-string formatting with CONFIG properties:
```python
# Before (hardcoded)
metadata_dir = pathlib.Path(temp_dir) / ".smartsim" / "metadata"

# After (CONFIG-based)
metadata_dir = pathlib.Path(temp_dir) / CONFIG.metadata_subdir
```

### Test Improvements
- **Reverted Parameterization**: Restored individual entity testing in `test_batch_symlink` and `test_symlink` for better test clarity
- **Consistent Imports**: Added CONFIG imports to all relevant test files
- **Path Flexibility**: Tests can now easily modify directory structures for different scenarios

## Testing

- ✅ All existing tests pass with CONFIG property updates
- ✅ Directory hierarchy correctly propagates changes (base → dragon → logs)
- ✅ Metadata integration tests validate proper directory structure
- ✅ Dragon launcher tests work with new nested log directory structure
- ✅ Configuration tests verify expected property values
- ✅ Symlinking tests work with CONFIG-based paths
- ✅ Manifest metadata tests validate hierarchical structure

## Migration Notes

**For Developers:**
- Use `CONFIG.metadata_subdir` instead of hardcoding ".smartsim/metadata"
- Use `CONFIG.dragon_logs_subdir` instead of hardcoding ".smartsim/logs"  
- Use `CONFIG.smartsim_base_dir` for base directory references
- All directory properties automatically adapt to base directory changes

**For Users:**
- No breaking changes to public APIs
- Directory structure remains the same by default
- Enhanced consistency in file organization

## Backward Compatibility

- All existing directory paths remain unchanged by default
- Tests continue to work without modification for end users
- Public APIs maintain compatibility
- File organization is improved without breaking existing workflows

## Summary

This PR significantly improves SmartSim's codebase quality by:
1. **Removing deprecated functionality**: Telemetry, SmartDashboard, and indirect entrypoints
2. **Establishing robust directory management**: Centralized CONFIG-based path system
3. **Eliminating technical debt**: Removed 20+ hardcoded path references across 15+ files
4. **Improving code organization**: Hierarchical directory structure with proper dependency relationships
5. **Enhancing maintainability**: Single source of truth for all directory paths
6. **Future-proofing the codebase**: Easy to extend with new directory requirements

The changes result in a cleaner, more maintainable codebase with improved consistency, better separation of concerns, and a solid foundation for future development.

Commit:
Remove telemetry functionality

- Remove TelemetryConfiguration classes and related code
- Remove telemetry monitor entrypoint and utilities
- Remove telemetry collectors and sinks
- Remove telemetry-related tests
- Remove watchdog dependency
- Simplify job entities and controller logic
- Remove telemetry configuration from config.py

This removes approximately 5,838 lines of telemetry-related code
while preserving core SmartSim functionality.

Commit:
Clean up remaining telemetry references and fix imports

- Remove telemetry_dir usage from controller.py batch job creation
- Clean up telemetry references in job.py comments and docstrings
- Remove telemetry-related properties from manifest.py
- Update serialize.py to remove telemetry directory and metadata references
- Remove telemetry_dir argument from indirect.py entrypoint and step.py launcher
- Update indirect tests to remove telemetry_dir parameter expectations
- Fix conftest.py to import JobEntity from correct location
- Clean up remaining telemetry comments and replace with generic logging

All telemetry code, configuration, tests, and documentation have now been
completely removed from the SmartSim codebase.

Commit:
Remove final telemetry references from codebase

- Clean up remaining telemetry references in job.py comments
- Simplify step.py proxy decorator to always use direct launch
- Remove telemetry.disable() call from CLI validate.py
- Simplify dragon backend cooldown period configuration
- Remove unused get_config import from dragon backend

All telemetry code has been completely removed from SmartSim.
The codebase now works without any telemetry dependencies or references.

Commit:
Fix indirect tests after telemetry removal

- Replace CONFIG.telemetry_subdir references with 'status' directory
- Remove telemetry event tracking from test_process_failure and test_complete_process
- Simplify tests to focus on actual process execution rather than telemetry events
- All indirect tests now pass without telemetry dependencies

Tests now verify core functionality without relying on removed telemetry system.

Commit:
Remove SmartDashboard integration and references

- Remove dashboard CLI plugin and all associated functionality
- Remove SmartDashboard documentation file (smartdashboard.rst)
- Update documentation index to remove SmartDashboard section
- Clean up ReadTheDocs configuration to remove dashboard dependency
- Update Docker files to remove SmartDashboard installation
- Remove dashboard-related tests and update plugin tests
- Update changelog to document SmartDashboard removal as breaking change
- Remove SmartDashboard changelog section

SmartSim now operates independently without SmartDashboard integration.
The core monitoring and logging functionality is preserved through
SmartSim's existing logging infrastructure.

Commit:
Fix mypy type annotation errors in CLI plugin system

- Add proper type annotation for empty plugins tuple in plugin.py
- Add explicit type annotation for plugin_items in cli.py
- All mypy checks now pass successfully

Commit:
Fix remaining test failures and clean up telemetry remnants

- Remove telemetry-related test functions from test_experiment.py
- Fix status_dir metadata by setting it to .smartsim subdirectory
- Fix controller test expecting removed exp_path parameter
- All tests now pass and mypy is clean

Commit:
Clean up remaining telemetry references in test files

- Remove telemetry-related test functions from test_config.py and test_serialize.py
- Remove telemetry fixtures and references from test_logs.py and conftest.py
- Update manifest_json fixture to use simple path instead of telemetry_subdir
- All tests now pass without telemetry dependencies

Commit:
make style

Commit:
Fix test expectations for new output file structure

- Updated test_output_files.py to match simplified .smartsim directory structure
- Updated test_symlinking.py to use new output file paths
- Fixed controller to use absolute paths for status directories
- Implemented historical file preservation with timestamps
- Updated batch job tests to use correct entity relationships
- Modified symlink_error test to match new auto-creating behavior

All core telemetry removal is complete with only output redirection issues remaining.

Commit:
Fix all lint errors to unblock CI/CD

- Remove unused imports (CONFIG, subprocess, sys, pathlib, get_ts_ms, encode_cmd, UnproxyableStepError)
- Fix line length issues in indirect.py and job.py
- Remove unreachable code after return statements
- Remove unused variables (start_rc, status_dir, is_dragon)
- Fix import-outside-toplevel issue with time module in controller.py
- Add pylint disable comment for unused argument raw_experiment
- Remove unnecessary pass statement and simplify docstring

All lint checks now pass with 10.00/10 rating.

Commit:
Last fixes

Commit:
Fix

Commit:
Indirect timestamp functionality added back

Commit:
Remove indirect entrypoint and corresponding tests

- Delete smartsim/_core/entrypoints/indirect.py
- Delete tests/test_indirect.py
- Update step.py comment to remove references to indirect launching
- Clean up cached files and mypy cache for removed modules
- Verified all tests pass and no type errors remain

Commit:
Remove spurious files

Commit:
Fix test failures and clean up remaining telemetry references

- Fix KeyError for status directory in batch job steps by setting status_dir in _create_batch_job_step
- Remove test_orc_telemetry test that referenced deleted telemetry functionality
- Remove remaining telemetry environment variable settings from dragon and pals tests
- Update line formatting for better lint compliance
- All originally failing tests now pass

Commit:
Remove lingering files

Commit:
Fix lingering output files in test_symlinking and test_output_files

- Enhanced symlink_output_files to auto-create parent directories
- Fixed path handling for entities with sub-entities (Orchestrator/Ensemble)
- Ensured all tests use proper test directories instead of repo root
- Removed unused CONFIG imports
- All tests now pass without creating lingering files in repo root

Commit:
Refine changelog

Commit:
Remove unused error class

Commit:
Remove proxyable command

Commit:
Restore step information in dictified model

Commit:
Fix serialize calls

Commit:
Remove unused telemetry fixtures from conftest.py

- Remove MockSink class and mock_sink fixture
- Remove mock_con, mock_mem, mock_redis, and mock_entity fixtures
- Remove MockCollectorEntityFunc protocol
- Clean up unused imports (asyncio, DragonLauncher, JobEntity)
- Improves pylint score from 9.56 to 9.67

Commit:
Remove defensive mkdirs

Commit:
Revert symlinking test

Commit:
Remove obsolete lines

Commit:
Implement consistent metadata directory pattern

- Add CONFIG.metadata_subdir property following established pattern
- Refactor controller to use consistent .smartsim/metadata base path
- Replace timestamped run_dir with metadata_dir/run_timestamp structure
- Update all method signatures: run_dir -> metadata_dir parameters
- Preserve historical log functionality with timestamped subdirectories
- Update tests to work with new metadata directory pattern
- Add test coverage for new CONFIG.metadata_subdir property

Addresses reviewer feedback for consistent directory structure
while maintaining backward compatibility and historical logs.

Commit:
Removed unused completion status logic

Commit:
Reinstate metadata_dir

Commit:
Fix style

Commit:
Fix lint

Commit:
Fix metatdata_dir occurrences

Commit:
Make metadata_dir mandatory in _create_batch_job_step

- Changed _create_batch_job_step to require metadata_dir as mandatory parameter
- Removed optional parameter and associated conditional logic
- Updated docstring to reflect mandatory parameter
- Updated test_controller.py to provide metadata_dir argument
- All lint and mypy checks pass
- All existing tests continue to pass

Addresses reviewer feedback about unnecessary optional parameter checks.

Commit:
Make metadata_dir mandatory in _create_job_step

- Changed parameter from Optional[pathlib.Path] = None to pathlib.Path
- Removed conditional logic for handling None metadata_dir
- Updated docstring to remove 'Optional' from parameter description
- Simplified implementation by always using provided metadata_dir
- All callers already provide metadata_dir, making this change safe
- Maintains consistency with _create_batch_job_step changes

Commit:
Refactor metadata directory management to use LaunchedManifestBuilder

- Added exp_metadata_subdirectory and run_metadata_subdirectory properties to LaunchedManifestBuilder
- These replace the old exp_telemetry_subdirectory and run_telemetry_subdirectory concepts
- LaunchedManifestBuilder now manages timestamp creation and directory structure
- Controller _launch method now uses manifest_builder.run_metadata_subdirectory instead of creating metadata_dir locally
- Added time import to manifest.py for timestamp generation
- Maintains consistent timestamp across the entire launch session
- Addresses reviewer feedback to use LaunchedManifestBuilder for metadata directory management

Commit:
Remove unused pylint pragma

Commit:
Remove redundant mkdirs

Commit:
Revert _launch_orchestrator signature to remove metadata_dir parameter

- _launch_orchestrator method no longer takes metadata_dir as a parameter
- Instead it gets the metadata directory internally from manifest_builder.run_metadata_subdirectory
- This restores the original cleaner method signature
- _launch method still creates metadata_dir locally since other methods need it
- All tests pass and mypy/lint checks are clean

Commit:
Restore entity-type-specific metadata directories

- Added get_entity_metadata_subdirectory() method to LaunchedManifestBuilder
- Each entity type (model, ensemble, database) now gets its own metadata subdirectory
- Structure: .smartsim/metadata/run_{timestamp}/{entity_type}/
- Updated controller to use type-specific directories:
  - Models use model_metadata_dir
  - Ensembles use ensemble_metadata_dir
  - Databases use database_metadata_dir
- This restores the original telemetry behavior but with new metadata naming
- All integrity checks pass (mypy, lint, tests)
- Directory structure verified to work correctly

Commit:
Fix controller

Commit:
Add tests

Commit:
make style

Commit:
Remove useless mkdirs

Commit:
Udpate serialization path

Commit:
Fix tests

Commit:
make style

Commit:
Update metadata_dir structure

Commit:
Update changelog

Commit:
make style

Commit:
Revert symlinking test parameterization

Commit:
Revert test_symlink parameterization

Commit:
Use type, not stringified type

Commit:
Fix test

Commit:
Remove hard-coded .smartsim occurrences

Commit:
Update dragon log dir

Commit:
Update changelog

al-rigazzi reviewed (commented):
In general, after some assisted modifications, I believe the structure now looks reasonable. Note that the dashboard has already been deprecated in its own repo.

MattToast commented on tests/test_controller_metadata_usage.py:
```diff
@@ -0,0 +1,169 @@
+
+    def teardown_method(self):
+        """Clean up test fixtures"""
+        import shutil
```
Nit: we are importing `shutil` outside of the global scope? Is there a reason for this?

MattToast commented on smartsim/_core/_cli/cli.py:
```diff
@@ -62,7 +62,9 @@ def __init__(self, menu: t.List[MenuItemConfig]) -> None:
 
         self.register_menu_items(menu)
-        self.register_menu_items([plugin() for plugin in plugins])
+        # Register plugin menu items (currently empty since all plugins were removed)
+        plugin_items: t.List[MenuItemConfig] = [plugin() for plugin in plugins]
+        self.register_menu_items(plugin_items)
```
Nit: Now that we require Python >= 3.9, prefer `list[T]` over the (now deprecated) `typing.List[T]`.

```suggestion
        # Register plugin menu items (currently empty since all plugins were removed)
        plugin_items: list[MenuItemConfig] = [plugin() for plugin in plugins]
        self.register_menu_items(plugin_items)
```

Or, better yet, drop the type-hint entirely and let the checker figure it out!

```suggestion
        # Register plugin menu items (currently empty since all plugins were removed)
        plugin_items = [plugin() for plugin in plugins]
        self.register_menu_items(plugin_items)
```

MattToast commented on smartsim/_core/control/job.py:
```diff
@@ -35,8 +35,8 @@
-    monitor. These keys are not guaranteed to be unique across experiments,
+    """A helper class for creating unique lookup keys within a job manager.
+    These keys are not guaranteed to be unique across experiments,
     only within an experiment (due to process ID re-use by the OS)"""
```
If you want to do some more clean-up work (and iirc), `_JobKey` is only ever used as a return value for `JobEntity.key`, and from a fairly naive grep, I could find any references to that attribute.

You should be free to remove both this class and `JobEntity.key`!

al-rigazzi replied:
Done

MattToast commented on smartsim/_core/control/job.py:
```diff
@@ -147,22 +122,22 @@ def _map_standard_metadata(
+        entity.metadata_dir = str(metadata_dir)
 
     @classmethod
     def from_manifest(
```
Similar note here: `JobEntity.from_manifest` was an alternate constructor for making `JobEntity` instances from information provided in the manifest file for the Telemetry Monitor. Since the TM no longer exists, this constructor appears to be unused!

MattToast replied:
And then by removing this constructor, you can remove both `JobEntity._map_standard_metadata` and `JobEntity._map_db_metadata`!

MattToast replied:
And actually!! It doesn't look like we construct a `JobEntity` anywhere in the code base anymore. Naively removing `JobEntity` from the union types in `Job.__init__` and `JobManager.add_job` does not upset MyPy. Can we just remove `JobEntity` entirely??

al-rigazzi replied:
Done

MattToast commented on smartsim/_core/config/config.py:
```diff
@@ -272,24 +272,20 @@ def test_mpi(self) -> bool:  # pragma: no cover
-    def dragon_default_subdir(self) -> str:
-        return ".smartsim/dragon"
+    def metadata_subdir(self) -> str:
+        return f"{self.smartsim_base_dir}/metadata"
```
Nit: consider making these return `pathlib.Path`s just so that the type-checker can help catch "stupid mistakes" (e.g. passing a path to an arbitrary function that expects a string that is supposed to represent something else).

If, for whatever reason, that requires work too much work to fit in this PR and the return type needs to remain a string, consider replacing the manual path concatenation via f-str with `os.path.join` for better portability (even though windows isn't directly supported hahaha)

al-rigazzi replied:
Done

MattToast commented on smartsim/_core/control/manifest.py:
```diff
@@ -256,6 +249,9 @@ class LaunchedManifestBuilder(t.Generic[_T]):
     run_id: str = field(default_factory=_helpers.create_short_id_str)
+    _launch_timestamp: str = field(
+        default_factory=lambda: str(int(time.time() * 1000)), init=False
+    )
```
from what I can tell `run_id` and `_launch_timestamp` serve a very similar purpose (to determine which `SmartSimEntity`s were launched in which `Experiment.start` calls).

It also looks like `run_id` is only being dumped in the `manifest.json` file, but is not read anywhere. Is it strictly needed anymore if the Telemetry Monitor is no longer using that as input? Are we keeping it around just in case we need to bring something back?

al-rigazzi replied:
Nah, deleted `run_id`

MattToast commented on smartsim/_core/control/manifest.py:
```diff
@@ -197,17 +198,9 @@ class _LaunchedManifestMetadata(t.NamedTuple):
-
     @property
     def manifest_file_path(self) -> pathlib.Path:
-        return self.exp_telemetry_subdirectory / _serialize.MANIFEST_FILENAME
+        return pathlib.Path(self.exp_path) / _serialize.MANIFEST_FILENAME
```
Just sanity checking that this was intended:

This change will make the `manifest.json` file "public" being at the root of the experiment directory. 

In the past we wanted to keep this file "private" by hiding it in a hidden `{exp_path}/.smartsim/telemetry` directory (presumably because we did not want users to modify it, thus invalidating the dashboard output).

Now that the dashboard is not supported, we don't particularly have a reason to hide the `manifest.json` file, outside of _maybe_ some lingering correctness errors surrounding multiple experiments in the same driver script with the same `exp_path`.

Just wanted to make sure that it was intended that it end up in `{exp_path}` and not `{exp_path}/.smartsim/metadata` or similar!

MattToast replied:
Alternatively, if we no longer support the dashboard, there may be an argument to simply NOT generate the `manifest.json` file at all? Thoughts??

al-rigazzi replied:
I know. I was going to ask your opinion on this. I don't think we ever advertised the existence of `manifest.json` to users, so we may just drop it?

MattToast replied:
I definitely am going to elect to at least keep it "private" in `{exp_path}/.smartsim/...` just because of those lingering correctness issues I mentioned. E.g. this experiment
```py
from smartsim import Experiment

exp_1 = Experiment("my-exp-1", exp_path="/tmp/some-dir")
rs_1 = exp_1.create_run_settings("echo", exe_args=["Hello", "from", "Experiment", "1"])
m_1 = exp.create_model("hello-1", rs_1)

exp_2 = Experiment("my-exp-2", exp_path="/tmp/some-dir")
rs_2 = exp_1.create_run_settings("echo", exe_args=["Hello", "from", "Experiment", "2"])
m_2 = exp.create_model("hello-2", rs_2)

exp_1.start(m_1)
exp_2.start(m_2)
```
will generate a SINGLE `/tmp/some-dir/manifest.json` file and it will report that BOTH models `m_1` and `m_2` were launched by Experiment 1. This can be particularly confusing if `exp_1` and `exp_2` were actually using _different launchers_. AND this on top of known limitations with the `manifest.json` fromat we chose (MPMD jobs for e.g. cannot be properly JSON-ified)

That said, personally, I would be okay with just dropping it entirely!! As far as I can remember, it was only ever meant as a stop-gap solution for the Telemetry Monitor until we designed something more sophisticated. It seems unlikely that we will have the manpower required to address these correctness issues, so if nothing is using the file internally AND we think that we should hide the file from end users, I see very little reason to make it in the first place!

al-rigazzi replied:
Well, let's delete them. They were a good stopgap solution, btw.

MattToast reviewed (changes requested):
Still making my way through the rest of this PR, but I have some initial thoughts I wanted to throw your way instead of stalling any longer 😅 

Feel free to lmk what you think!!

Commit:
Update smartsim/_core/_cli/cli.py

Co-authored-by: Matt Drozt <matthew.drozt@gmail.com>

Commit:
Address MattToast's code review feedback (items 1-3)

Implement the following improvements from PR #789 code review:

1. Fix import style: Move shutil import to module level in test_controller_metadata_usage.py
   - Relocate shutil import from method to top-level imports per Python best practices

2. Remove unused JobEntity code: Complete cleanup of JobEntity ecosystem
   - Remove JobEntity class and _JobKey class from job.py
   - Remove JobEntity imports and isinstance checks from jobmanager.py
   - Simplify Job type annotations to use actual SmartSim entities only
   - Eliminate telemetry-related legacy code that's no longer needed

3. Enhance CONFIG with Path objects: Improve type safety for directory paths
   - Update smartsim_base_dir, dragon_default_subdir, dragon_logs_subdir, metadata_subdir
     to return pathlib.Path objects instead of strings
   - Maintain backward compatibility with os.path.join and string operations
   - Update test expectations to validate Path object behavior

All changes tested and verified:
- Import style follows Python conventions
- JobEntity references completely removed from codebase
- Path objects provide enhanced type safety while preserving compatibility
- All existing tests pass with new Path-based CONFIG properties

Commit:
Remove unused run_id from manifest system

Address MattToast's feedback about removing run_id which was used for
telemetry tracking but is no longer needed after telemetry removal.

Changes:
- Remove run_id field from _LaunchedManifestMetadata NamedTuple
- Remove run_id parameter from LaunchedManifestBuilder constructor
- Remove run_id from serialized manifest.json output
- Update all test files to remove run_id parameters
- Update test expectations to use timestamp for uniqueness instead

The manifest system now uses timestamp for run identification instead
of the UUID-based run_id, simplifying the codebase after telemetry removal.

Commit:
make style

Commit:
Merge branch 'develop' of https://github.com/CrayLabs/SmartSim into drop_telemetry

Commit:
Minor changes to headers

Commit:
Update copyright

Commit:
Remove LaunchedManifest classes and clean up telemetry code

- Remove LaunchedManifest, _LaunchedManifestMetadata, and LaunchedManifestBuilder classes
- Simplify serialize.py by removing orphaned telemetry functions (80% reduction)
- Update controller.py to remove LaunchedManifest dependencies and phantom method call
- Clean up all test files to remove LaunchedManifest references
- Delete tests/test_serialize.py as it only tested removed functionality
- Maintain core Manifest class functionality for entity organization
- Achieve 10.00/10 linting score across all modified files

Commit:
Fix orchestrator checkpoint saving

- Restore missing _save_orchestrator() call in _launch_orchestrator_simple()
- This was accidentally removed during LaunchedManifest cleanup
- Fixes test_dbnode.py::test_hosts which requires checkpoint file for reconnection
- Maintains 10.00/10 linting score

Commit:
Changelog refinement

Commit:
Fix database host setup in orchestrator launch

- Restore missing _jobs.set_db_hosts(orchestrator) call in _launch_orchestrator_simple()
- This was accidentally removed during LaunchedManifest cleanup
- Fixes IndexError in db_is_active() where hosts list was empty
- Resolves backend ML model test failures (test_dbmodel.py, test_dbscript.py)
- Database addresses now properly populated for entity launches
- Maintains 10.00/10 linting score

Commit:
Fix metadata directory uniqueness for multiple model runs

- Add timestamp-based unique metadata directories for each launch
- Import get_ts_ms helper function from utils.helpers
- Modify ensemble and model metadata directory paths to include launch timestamp
- Ensures each experiment launch gets unique metadata directories
- Fixes test_output_files.py::test_mutated_model_output
- Prevents output file overwrites when same model is run multiple times
- Historical output files now properly preserved across multiple runs
- Maintains 10.00/10 linting score

Commit:
Move TStepLaunchMetaData to controller_utils.py and remove serialize.py

- Move TStepLaunchMetaData type definition from serialize.py to controller_utils.py
- Remove unused smartsim/_core/utils/serialize.py file entirely
- Add pathlib.Path import to controller_utils.py for type definition
- Remove TYPE_CHECKING import that was only used for the moved type
- Complete final cleanup of telemetry-related serialization code
- All functionality preserved and tests still pass

Commit:
Remove unused code

Commit:
Remove unused test file

Commit:
Revert wrong indentation

Commit:
Remove comments

Commit:
Remove comments

Commit:
Make style

MattToast commented on smartsim/_core/launcher/dragon/dragonBackend.py:
```diff
@@ -177,12 +176,7 @@ def __init__(self, pid: int) -> None:
-            if smartsim_config.telemetry_enabled
-            else 5
-        )
+        self._cooldown_period = 5
```
Nit, not to be addressed in this PR: Do we not have an environment variable or something to let users' configure the cool down period? Should be add this as a field to `Config`, or as a parameter to `__init__` or something?

al-rigazzi replied:
Right, but we may be removing the whole ZMQ thing in the future, so maybe we won't need this anymore.

MattToast replied:
Ahhh! In that case, this might be something not worth being addressed at all lol 😂

MattToast commented on smartsim/_core/launcher/step/step.py:
```diff
@@ -27,20 +27,17 @@
 
 from ....log import get_logger
 from ....settings.base import RunSettings, SettingsBase
-from ...utils.helpers import encode_cmd, get_base_36_repr
```
Nit: If you want to clean up _EVEN MORE_, it looks like `encode_cmd` and `decode_cmd` defined in `smartsim._core.utils.helpers`, but not called anywhere outside of test files

al-rigazzi replied:
Done and done!

MattToast commented on tests/test_controller.py:
```diff
@@ -69,7 +70,7 @@ def test_controller_batch_step_creation_preserves_entity_order(collection, monke
-    _, steps = controller._create_batch_job_step(
-        collection, pathlib.Path("mock/exp/path")
-    )
+    # Create a metadata directory for the test
+    metadata_dir = pathlib.Path("/tmp") / CONFIG.metadata_subdir
+    _, steps = controller._create_batch_job_step(collection, metadata_dir)
```
Nit: should we consider making this a path based off of the `test_dir` fixture, just so that any artifacts that are created during the test are easier to be found/investigated in the case of a failure?

MattToast commented on tests/test_output_files.py:
```diff
@@ -106,10 +106,12 @@ def test_mutated_model_output(test_dir):
-    step = controller._create_job_step(model, status_dir)
-    expected_out_path = status_dir / model.name / (model.name + ".out")
-    expected_err_path = status_dir / model.name / (model.name + ".err")
+    model.path = test_dir
```
Nit: Not really a fan of modifying attributes of globally available instances of `Model`s just because it means that this test CAN leak state in future. If we need to set this path, can we make this `Model` a fixture, or use `monkeypatch.setattr` or something?

al-rigazzi replied:
Should be fixed now?

MattToast commented on tests/test_symlinking.py:
```diff
@@ -65,7 +65,7 @@ def test_symlink(test_dir, entity):
     """Test symlinking historical output files"""
     entity.path = test_dir
     if entity.type == Ensemble:
-        for member in ens.models:
+        for member in entity.models:
```
Nice catch!!

MattToast commented on tests/test_symlinking.py:
```diff
@@ -100,19 +104,51 @@ def test_batch_symlink(entity, test_dir):
+    # For entities with sub-entities (like Orchestrator), set their paths too
+    if hasattr(entity, "entities"):
+        for sub_entity in entity.entities:
+            sub_entity.path = test_dir
```
Ditto previous complaint about this _technically_ leaking test state by writing to global variables (I know that we do this elsewhere in this module, e.g. line 106, don't worry it bothers me there too hahaha). Thoughts on making these entities fixtures and/or using `monkeypatch.setattr`??

al-rigazzi replied:
You know me, I love monkeys.

MattToast commented on tests/test_symlinking.py:
```diff
@@ -100,19 +104,51 @@ def test_batch_symlink(entity, test_dir):
+    else:
+        # For _AnonymousBatchJob (single model)
+        substep = substeps[0]
+        slurm_controller.symlink_output_files(substep, entity)
```
Just sanity checking, is this block reachable? In theory an `_AnonymousBatchJob` has an `entities` attribute that is set to a list of length 1 containing the single model at `__init__` time. If my mental compiler is correct, I think this block is unreachable and free to remove!

MattToast commented on tests/test_symlinking.py:
```diff
@@ -100,19 +104,51 @@ def test_batch_symlink(entity, test_dir):
+    # For batch entities, we need to call symlink_output_files correctly
+    # Based on how the controller does it, we should pass the individual entities
+    if hasattr(entity, "entities") and len(substeps) > 0:
+        # Just test the first substep and entity pair
```
Not wrong, just curious: Any reason we only test the first substep? Could we test them all??

al-rigazzi replied:
Nope, let's test them all. Also removed the defensive if.

MattToast commented on tests/test_metadata_integration.py:
```diff
@@ -0,0 +1,348 @@
+            ), f"Database metadata directory should not exist: {database_dir}"
+
+            # Clean up
+            exp.stop(model)
```
Nit: Not strictly wrong, but do we need to `stop` the model if we already polled for its completion earlier in the test? Is this just to prevent breaking on a possible API change?

MattToast commented on tests/test_metadata_integration.py:
```diff
@@ -0,0 +1,348 @@
+
+            assert metadata_dir.exists(), "Metadata directory should exist"
+
+            # Check for run-specific subdirectories (may be 1 or 2 depending on timing)
+            run_dirs = [
+                d
+                for d in metadata_dir.iterdir()
+                if d.is_dir() and d.name.startswith("run_")
+            ]
+            assert (
+                len(run_dirs) >= 1
+            ), f"Should have at least one run directory, found: {run_dirs}"
```
I'm guessing this is because the orchestrator will may not spin up in time? It doesn't look like we do any verification of the orchestrator directory anyway in this test, can we just remove its launch entirely??

al-rigazzi replied:
Removed orc launch

MattToast commented on tests/test_metadata_integration.py:
```diff
@@ -0,0 +1,348 @@
+
+            exp1.start(model1, block=False)
+            exp1.poll(interval=1)
+            exp1.stop(model1)
```
functionally equivalent to
```suggestion
            exp1.start(model1, block=True)
```
? Is this just anticipating a possible future API break?

MattToast commented on tests/test_metadata_integration.py:
```diff
@@ -0,0 +1,348 @@
+
+            exp2.start(model2, block=False)
+            exp2.poll(interval=1)
+            exp2.stop(model2)
```
Ditto here
```suggestion
            exp2.start(model2, block=True)
```

MattToast commented on tests/test_metadata_integration.py:
```diff
@@ -0,0 +1,348 @@
+                "batch_ensemble",
+                run_settings=exp.create_run_settings("echo", ["batch_world"]),
+                replicas=2,
+            )
```
Just verifying this is intended: It looks like this test is supposed to be launching a model/ensemble with batch settings, but the local launcher is being used and no batch settings were assign to either

al-rigazzi replied:
Removed

MattToast reviewed (changes requested):
Couple of nits on tests and such, but otherwise looks about ready to go on my end! Thanks for all thorough clean-up effort!!

Commit:
Address simple part MattToast's comments

Commit:
Fixed lint

Commit:
A few commits

Commit:
Align metadata directories with new tests

Commit:
Update metadata integration test for single run dir

Commit:
Style

Commit:
Reviewer's comments

Commit:
Style

MattToast commented on tests/test_controller_metadata_usage.py:
```diff
@@ -0,0 +1,85 @@
+
+    entity_path = tmp_path / "workdir"
+    entity_path.mkdir()
+    entity = type("Entity", (), {"name": "sample", "path": str(entity_path)})
```
Quick sanity check here: This is creating a new _type_ `Entity` equivalent to
```py
class Entity:
    name = "sample"          # <- Note: this is a class level var
    path = str(entity_path)  # <-       ditto here

entity = Entity  # <- `entity` is now an alias for the _type_ `Entity`
```
so `isintance(entity, type) == True` and `isinstance(e(), object) == True != isinstance(e(), type)`

is this intended or did you mean something like this:
```py
@dataclasses.dataclass  # <- decorator to make class level vars instance vars
class Entity:
    name: str = "sample"
    path: str = str(entity_path)

entity = Entity()  # <- `entity` now an instance of `Entity`
```
? _Technically_ this doesn't _really_ matter as we only expect one instance of this class, but I'm not gonna give you a review without being overly pedantic 😅

MattToast commented on tests/test_metadata_integration.py:
```diff
@@ -0,0 +1,269 @@
+
+    def test_experiment_creates_correct_metadata_directory_structure_model_only(self):
+        """Test that launching only models creates the correct directory structure"""
+        with tempfile.TemporaryDirectory() as temp_dir:
```
Minor nit: Should this be using the `test_dir` fixture just so that artifacts are left over in the event of a failure for manual inspection?

MattToast commented on tests/test_metadata_integration.py:
```diff
@@ -0,0 +1,269 @@
+from smartsim.settings import RunSettings
+
+
+class TestMetadataDirectoryIntegration:
```
Minor nit: If there is no shared stand-up/tear-down logic in this test class, can we make all of these tests top level functions, just to match the style of the rest of the test suite?

MattToast commented on tests/test_metadata_integration.py:
```diff
@@ -0,0 +1,269 @@
+        self,
+    ):
+        """Test that launching only ensembles creates the correct directory structure"""
+        with tempfile.TemporaryDirectory() as temp_dir:
```
ditto: `test_dir` fixture ask here

MattToast replied:
applies for remaining tests in this class/file as well

MattToast commented on tests/test_metadata_integration.py:
```diff
@@ -0,0 +1,269 @@
+
+            # Start and wait for completion
+            exp.start(ensemble, block=False)
+            exp.poll(interval=1)
```
Style nit: equivalent to `exp.start(ensemble, block=True)`? Any reason for the separation?

MattToast commented on tests/test_metadata_integration.py:
```diff
@@ -0,0 +1,269 @@
+            ), f"Model metadata directory should exist: {model_dir}"
+            assert (
+                ensemble_dir.exists()
+            ), f"Ensemble metadata directory should exist: {ensemble_dir}"
```
Just checking because it was added in other tests: do we want this check here also?
```suggestion
            assert (
                model_dir.exists()
            ), f"Model metadata directory should exist: {model_dir}"
            assert (
                ensemble_dir.exists()
            ), f"Ensemble metadata directory should exist: {ensemble_dir}"
            assert (
                not database_dir.exists()
            ), f"Database metadata directory should not exist: {database_dir}"
```

MattToast commented on tests/test_metadata_integration.py:
```diff
@@ -0,0 +1,269 @@
+            )
+
+            exp.start(model, block=False)
+            exp.poll(interval=1)
```
Similar style nit here

MattToast commented on tests/test_metadata_integration.py:
```diff
@@ -0,0 +1,269 @@
+                    assert model_dir.is_dir()
+                    assert model_dir.stat().st_mode & 0o700
+
+            exp.stop(model)
```
Style nit: `model` is polled for earlier in this test, in theory this line is redundant
```suggestion
```

MattToast commented on tests/test_metadata_integration.py:
```diff
@@ -0,0 +1,269 @@
+                model_dir = run_dir / "model" / "test_model"
+                if model_dir.exists():
+                    assert model_dir.is_dir()
+                    assert model_dir.stat().st_mode & 0o700
```
If this `model.exisits()` condition is not met, I would argue that something has gone wrong. Thoughts on changing it to an `assert` instead?
```suggestion
                aseert model_dir.exists()
                assert model_dir.is_dir()
                assert model_dir.stat().st_mode & 0o700
```

MattToast commented on tests/test_metadata_integration.py:
```diff
@@ -0,0 +1,269 @@
+                for d in metadata_dir.iterdir()
+                if d.is_dir() and d.name.startswith("run_")
+            ]
+            if run_dirs:
```
Similar argument here, if the `run_dirs` is empty, I argue the test failed. Thoughts on making this an `assert` as well?
```py
assert len(run_dirs) == 1, "Incorrect number of run dirs created"
# TODO: Dedent everything in this `if` block
```

MattToast commented on tests/test_symlinking.py:
```diff
@@ -48,24 +48,45 @@
+def anon_batch_model_entity():
+    batch_model = Model(
+        "batch_test_model",
+        params={},
+        path="",
+        run_settings=batch_rs,
+        batch_settings=bs,
+    )
+    return _AnonymousBatchJob(batch_model)
 
 
 @pytest.mark.parametrize(
-    "entity",
-    [pytest.param(ens, id="ensemble"), pytest.param(model, id="model")],
+    "entity_fixture",
+    ["ensemble_entity", "model_entity"],
+    ids=["ensemble", "model"],
 )
-def test_symlink(test_dir, entity):
+def test_symlink(test_dir, request, entity_fixture):
     """Test symlinking historical output files"""
+    entity = request.getfixturevalue(entity_fixture)
     entity.path = test_dir
     if entity.type == Ensemble:
-        for member in ens.models:
+        for member in entity.models:
```
Not you're change, but this must have gotten past a different review. In theory `entity.type` should always return a `str`, meaning this block is unreachable. Do you think this should read
```py
if entity.type == "Ensemble":
    ...
```
or
```py
if isinstance(entity, Ensemble):
    ...
```
instead??

al-rigazzi replied:
Yes, I believe you are right, let me see if I can change it and make this reachable

al-rigazzi replied:
Done

MattToast commented on tests/test_symlinking.py:
```diff
@@ -48,24 +48,45 @@
-)
-anon_batch_model = _AnonymousBatchJob(batch_model)
+
+@pytest.fixture
+def model_entity():
+    return Model("test_model", params={}, path="", run_settings=rs)
+
+
+@pytest.fixture
+def ensemble_entity():
+    return Ensemble("ens", params={}, run_settings=rs, batch_settings=bs, replicas=3)
+
+
+@pytest.fixture
+def orchestrator_entity():
+    return Orchestrator(db_nodes=3, batch=True, launcher="slurm", run_command="srun")
```
LOVELY!! Thank you for adding!!

MattToast reviewed (approved):
Couple of small nits in the test suite, but nothing that I'm willing to hold this PR up over (I've already held it up to long lol 😅)

Commit:
Address reviewer's comments

Commit:
Style

Commit:
Update changelog

al-rigazzi commented:
@MattToast thanks for the review. I believe I addresseed your last comments and I will merge soon.
